### PR TITLE
Add ability to associate user group members with playmode(s)

### LINKED ISSUE #6681: Add 'new sort' by in BN pages by game mode they're nominated
I referred to this page : https://osu.ppy.sh/groups/28

It would be helpful in my opinion especially with new mappers when they searched bn for nominated.

### PR DESCRIPTION
**Pull request by nekodex**
aka: The ability to mark NAT/BN members as responsible for specific game modes

This is a prerequisite for the nomination changes, as discussed in #6689 and sets up the usergroup/playmode mapping.

Enabling playmode support is done at a per-group level, while the playmodes themselves are defined on the user's membership relation (`phpbb_user_group`) - this means that we can assign playmodes at a per-usergroup-per-user level. e.g. a user can be a NAT for osu! mode but also a BN for osu!+osu!taiko (not sure if this will ever happen, but it's supported)

This PR also has some visible changes:
- playmode(s) are now shown in the user group badge (if enabled for the group and the member has playmodes defined):
Screen Shot 2020-10-08 at 3 43 13 pm
- user badges have also been added to user cards (to make group membership more obvious):
Screen Shot 2020-10-08 at 3 46 02 pm
- ability added to filter the groups page by playmode has been added (if `has_playmodes ` is set for the group):
Screen Shot 2020-10-08 at 3 57 34 pm

resolves #6681

---

Things to be aware of when deploying:
- [x] check if migrations need to be done beforehand manually (mainly concerned about `phpbb_user_group`)
- [x] group cache needs to be cleared after setting `has_playmodes` in DB: run `app('groups')->resetCache()` via tinker, (this will need to be done again whenever `has_playmodes` is toggled on a group)
- [x] will need to construct/run some DB queries to set the relevant playmodes for current NAT/BN/etc members (update: constructed, just needs to be run after deploy)

### CODE REVIEW

**Comment by nekodex** [comment]
marking as blocked because I noticed some edge case issues

**Review by nanaya — CHANGES_REQUESTED** [review]

File: resources/assets/coffee/osu_common.coffee [hunk 1]
```diff
@@ -262,6 +262,18 @@
       window.scrollTo position[0], position[1]
 
 
+  playmodeToInt: (playmode) ->
```
Comment (nanaya) [inline resources/assets/coffee/osu_common.coffee:265]: not used?

File: resources/assets/lib/globals.d.ts [hunk 2]
```diff
@@ -62,12 +62,14 @@ interface OsuCommon {
   classWithModifiers: (baseName: string, modifiers?: string[]) => string;
   diffColour: (difficultyRating?: string | null) => React.CSSProperties;
   groupColour: (group?: GroupJSON) => React.CSSProperties;
+  intToPlaymode: (int: number) => string;
```
Comment (nanaya) [inline resources/assets/lib/globals.d.ts:65]: doesn't exist?

File: app/Models/User.php [hunk 3]
```diff
@@ -1373,19 +1373,34 @@ public function visibleGroups()
...
-                $group = app('groups')->byId($id);
+            foreach ($this->userGroups as $userGroup) {
+                $cachedGroup = app('groups')->byId($userGroup->group_id);
+                if (!$cachedGroup) {
+                    continue;
+                }
+
+                $group = clone $cachedGroup;
+
+                if ($group && $group->has_playmodes) {
```
Comment (nanaya) [inline app/Models/User.php:1385]: the first [truthy](https://www.php.net/manual/en/language.types.boolean.php#language.types.boolean.casting) check is rather useless. Similarly the `optional` wrapper below. Additionally there's no need to `clone` if `display_order` is `null`.

**Review by nanaya — CHANGES_REQUESTED** [review]

**Review by nanaya — CHANGES_REQUESTED** [review]

File: resources/docs/source/includes/_structures/group.md [hunk 4]
```diff
@@ -1,10 +1,18 @@
...
+short_name      | string | Short name of the group for display.
+description     | string | |
+colour          | string | |
+
+
+If [Group](#group) is returned as part of a [User](#user) object, it may also contain:
+
+Field           | Type     | Description
+----------------|----------|------------------------------------------------------------
+playmodes       | string[] | [GameModes](#gamemode) which the member is responsible for, e.g. in the case of BN/NAT
```
[the remark this answers is not in the document]
Comment (nekodex) [reply resources/docs/source/includes/_structures/group.md:18]: attribute doesn't exist at all unless it's inside user, which is why I listed it separately... but yeah, also nullable

[the remark this answers is not in the document]
Comment (nanaya) [reply resources/docs/source/includes/_structures/group.md:18]: the transformer doesn't imply that though? :thinking:

**Review by nanaya — APPROVED** [review]

**Comment by nanaya** [comment]
`phpbb_user_group` seems to be only about 500MB so it should be fine even run from migration.

**Comment by peppy** [comment]
Remaining two steps will happen after deploy.

**Pull request merged**